# Methods

## Datahub: UCSC Xena

The UCSC Xena resources were accessed using the `UCSCXenaTools` package [@Li2024]. Specifically, tumour-normal comparisons were based on the Toil workflow [@Vivian2017] *"where TCGA, TARGET, and GTEx samples are re-analyzed by the same RNA-seq pipeline. This pipeline involved re-aligning the reads to hg38 genome and calling gene expression using RSEM and Kallisto methods. Because all samples are processed using a uniform bioinformatic pipeline, batch effects due to different computational processing is eliminated. Note that the samples from this study have only undergone per-sample normalization...There are four gene expression datasets in this study. Two are normalized using with-in sample methods. The 'RSEM norm__count' dataset is normalized by the upper quartile method, the 'RSEM expected__count (DESeq2 standardized)' dataset is by DESeq2 normalization."* 

We retrieved the following Toil datasets:

* TCGA-GTEx-TARGET-gene-exp-counts.deseq2-normalized.log2
* TcgaTargetGtex_RSEM_Hugo_norm_count
* TcgaTargetGtex_expected_count
* TcgaTargetGtex_rsem_isoform_tpm
* TcgaTargetGTEX_phenotype.txt
* TCGA_survival_data
* TCGA_GTEX_category.txt
* TCGA_TARGET_clin.subtype_hist.tab
* Survival_SupplementalTable_S1_20171025_xena_sp
* EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.xena
* broad.mit.edu_PANCAN_Genome_Wide_SNP_6_whitelisted.gene.xena

Samples annotated as "Solid Tissue Normal" were initially re-assigned to a universal "Negative control" category. However, expression levels of tumour-specific markers (e.g. NECTIN4 in Bladder) are suspiciously high in this mixed-tissue category, suggestive of sample contamination or sample swaps. As such, this category was subsequently dropped from this report, and further analysis (e.g. PCA) will be required to clarify sample identities.

To assign breast cancers into TNBC, ER+ and HER2+ sub-categories, the clinical data for breast cancer was complemented with histologic annotations as described [@Thennavan2021].

Differentially expressed genes/transcripts were identified using the `EMDomics` package [@Nabavi2016]. This uses the Earth mover's distance to measure the overall difference between the distributions of a gene's expression in two classes of samples. This technique is suited to samples that exhibit both significant intra- and inter-class heterogeneity. In this case, tumour samples were compared either to the *combined* distribution of all healthy tissues or to the specific healthy distribution that has the highest median expression. To focus efforts on expressed genes, a minimum tumour expression of log2(RSEM expected_count + 1) > 1 was applied to trigger comparison. To identify the top 10 targets that consistently rank highly in these analyses, we used weighted rank aggregation on the ordered target list for each indication using the Cross-Entropy Monte Carlo algorithm and Kendalls tau distance in the `RankAggreg` package. Weights were assigned according to the mortality data found in the WHO Global Cancer Observatory [GCO](https://gco.iarc.who.int). This data was manually retrieved from the `Cancer Today` section: *"In brief, the national estimates are built up from the best available sources of cancer incidence and mortality data within each country, and their validity depends on the degree of representativeness and quality of the source information. The methods used to compile the 2022 estimates are largely based on those developed previously with an emphasis on the use of short-term predictions and the use of modeled mortality-to-incidence ratios, where applicable. The estimates are available in the GCO for 36 cancer types, including nonmelanoma skin cancer (NMSC) (International Classification of Diseases, Tenth Edition [ICD-10] code C44, excluding basal-cell carcinomas). Together with all cancers combined, cancer-specific estimates are provided for 185 countries or territories worldwide by sex and 18 age groups (ages birth to 4 years, 5–9 years, …, 80–84 years, 85 years and older)."* [@Bray2024; @Filho2025]



## Datahub: Precedented targets

This list was generated by manual literature review, and augmenting this with gene lists from two ADC atlases [@Fang2024; @Kathad2024].










